Wilderness area contrib

titeuf87 opened the pull request:
#### Brief overview of PR changes/additions
This adds a new contrib that adds a wilderness area.
The idea comes from Griatch from the [mailing list](https://groups.google.com/forum/#!searchin/evennia/ocean%7Csort:relevance/evennia/MGI6VYwMP2g/CTuyIksCDAAJ)

#### Motivation for adding to Evennia
This can be used by others so they have something to base their work on without having to reinvent the wheel.

#### Other info (issues closed, discussion etc)
I hope that the comments/docstrings are clear!

Griatch commented on evennia/contrib/wilderness.py:
```diff
@@ -0,0 +1,490 @@
+            self.wilderness.move_obj(moved_obj, location)
+        else:
+            # This object wasn't in the wilderness yet. Let's add it.
+            itemlocs[moved_obj] = self.ndb.active_location
```
While I can see later in the code that self.ndb.active_location is set by the wilderness script, this is the first time we see this property appear, so it would help readability to mention where it comes from, or even mention this in the class docstring.

Griatch commented on evennia/contrib/wilderness.py:
```diff
@@ -0,0 +1,490 @@
+                item.location = None
+        # And every obj matching the new location will get its location set to
+        # this room
+        for item, location in self.wilderness.db.itemlocations.iteritems():
```
The performance of this will scale linearly (O(N)) with the number of the objects in the wilderness. Which may be fine; for a larger set one could consider using Tags for the coordinates. Those scale as O(1) due to database indexing and would eventually be considerably more effective. Not something for this PR, but maybe worth to make a comment on for those having bigger performance requirements.

Griatch commented on evennia/contrib/wilderness.py:
```diff
@@ -0,0 +1,490 @@
+        # Key: object, Value: (x, y)
+        self.db.itemlocations = {}
+
+        # Store the rooms that are used as views into the wilderness
```
Does this mean you can insert "fixed" rooms into the dynamic grid too (like a dungeon entrance or town)? Considering new "dynamic" rooms are only taken from the `unused_rooms` list, this sounds like an easy additional feature to add, but would require a helper method to specify the fixed room and at which coordinate it should appear.

titeuf87 replied:
That's a good idea! Currently that's not possible as the rooms would be moved away to storage when they're not needed anymore. But adding a "keep" flag could fix that.

Griatch commented on evennia/contrib/wilderness.py:
```diff
@@ -0,0 +1,490 @@
+    return (x, y)
+
+
+class WildernessScript(DefaultScript):
```
This requires a docstring detailing what the purpose of it is.

Griatch commented on evennia/contrib/wilderness.py:
```diff
@@ -0,0 +1,490 @@
+
+    The defaults, while useable, are meant to be customised. When creating a new
+    wilderness map it is possible to give a "map provider": this is a python
+    object that is smart enough to create the map.
```
You should give a brief example of a Map Provider here, especially which methods the wilderness system expects it to have.

Griatch commented on evennia/contrib/wilderness.py:
```diff
@@ -0,0 +1,490 @@
+    This contrib does not provide any commands. Instead the @py command can be
+    used.
+
+    A wilderness map needs to created first. There can be different maps, all
```
It would be useful with a very brief example of creating such a map, here or at the end of this docstring.

titeuf87 commented:
I think I fixed all various docstrings/style issues. 
Added an example too on how to customise it.
Changed location to coordinates (which is indeed clearer)

Let me know if I missed something!

Griatch commented on evennia/contrib/wilderness.py:
```diff
@@ -0,0 +1,723 @@
+    """
+    Default Wilderness Map provider.
+
+    This is a simple provider that just creates an infite large grid area.
```
infite -> infinite

Griatch commented on evennia/contrib/wilderness.py:
```diff
@@ -0,0 +1,723 @@
+
+        return True
+
+    def get_location_name(self, coordinates):
```
Only a get_location_name method seems very limited for the kind of use this will see. The average developer will have to at least overload the description per-room, based on the coordinate. Would it not be suitable to also supply something like `at_select(self, coordinates, caller, room)` that gets called with the room object and the character moving when the room is being prepared and allows the user to assign data to the room based on its coordinate?

Griatch commented on evennia/contrib/wilderness.py:
```diff
@@ -0,0 +1,723 @@
+        return True
+
+
+class WildernessMapProvider(object):
```
Since this is sort of a default input to the other entities, it should probably be at the very end, after the WildernessScript, which is not intended to be modified as far as I understand.

Griatch commented:
I tested the example commands and could move around on the default infinite wilderness just fine! I added some usability comments.

titeuf87 commented:
That `at_select` was a good idea. I renamed it to `at_prepare_room` though to make it clearer what it does. Also fixed your other comments.

Griatch commented:
I merged with a rebase. One thing struck me though when looking at this. There are dicts for storing the position of objects in the wilderness, but I don't see any examples for a developer to actually manage objects and which methods they would use.

titeuf87 commented:
What exactly do you mean by managing the objects inside the wilderness?

The only issue I can think of right now is that while you can add any object in the wilderness (using the `enter_wilderness`, it is not actually possible to remove random objects back out of the wilderness unless they just happen to be at the same coordinates than a player is standing in.

Griatch commented:
Aha, you use enter_wilderness to put stuff in the wilderness. That makes sense in retrospect, but it was not obvious to me from the doc.
